drivers/platform/maxim/max32670/maxim_gpio.c

```diff
@@ -0,0 +1,343 @@
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include <stdio.h>
+#include <errno.h>
+#include <stdlib.h>
+#include "no_os_gpio.h"
+#include "no_os_irq.h"
+#include "no_os_util.h"
+#include "gpio.h"
+#include "gpio_regs.h"
+#include "maxim_gpio.h"
+#include "maxim_irq.h"
+#include "max32670.h"
+#include "mxc_errors.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Set/Clear GPIO enable register.
+ * @param regs - The GPIO register structure.
+ * @param mask - GPIO pins mask.
+ * @param is_enabled - Enable status.
+ * @return none.
+ */
+static void set_enable(mxc_gpio_regs_t *regs, uint32_t mask, uint8_t is_enabled)
+{
+	if (is_enabled)
+		regs->en0 |= mask;
+	else
+		regs->en0 &= ~mask;
+}
+
+/**
+ * @brief Obtain the GPIO descriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - GPIO initialization parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t max_gpio_get(struct no_os_gpio_desc **desc,
+		     const struct no_os_gpio_init_param *param)
+{
+	int32_t ret;
+	mxc_gpio_cfg_t *g_cfg;
+	struct max_gpio_init_param *pextra;
+	uint32_t m_pad, m_func;
+	struct no_os_gpio_desc *descriptor;
+
+	if (!param || !param->extra || param->number >= N_PINS)
+		return -EINVAL;
+
+	descriptor = calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	g_cfg = calloc(1, sizeof(*g_cfg));
+	if (!g_cfg) {
+		ret = -ENOMEM;
+		goto free_descriptor;
+	}
+
+	pextra = param->extra;
+
+	switch(param->pull) {
```

Comment:
Perhaps it would be better if we also add the `NO_OS_PULL_UP_WEAK` and `NO_OS_PULL_DOWN_WEAK` as fallthrough cases to `NO_OS_PULL_UP` and `NO_OS_PULL_DOWN` if the max32670 doesn't support setting more than one resistance value for the pull ups.

Reply:
Make sense.